db/db_test2.cc

```diff
@@ -4271,6 +4280,62 @@ TEST_F(DBTest2, TraceAndReplay) {
   ASSERT_OK(db2->Get(ro, handles[1], "rocksdb", &value));
   ASSERT_EQ("rocks", value);
 
+  // Re-replay should fail with Status::Incomplete() if Prepare()
+  // was not called.
+  ASSERT_TRUE(replayer->Replay().IsIncomplete());
+
+  // Re-replay using 2 threads, 2x speed.
+  ASSERT_OK(replayer->Prepare());
+  ASSERT_OK(replayer->Replay(ReplayOptions(2, 2.0)));
+
+  // Re-replay using 2 threads, 1/2 speed.
+  ASSERT_OK(replayer->Prepare());
+  ASSERT_OK(replayer->Replay(ReplayOptions(2, 0.5)));
+
+  // Manual replay
+  // Iteration 1: Manul replay should succeed;
+  // Iteration 2: Manul replay should fail with no Prepare();
+  // Iteration 3: Manul replay again should succeed with Prepare();
+  for (int i = 1; i <= 3; i++) {
+    // Do not Prepare() for the 2nd iteration,
+    if (i != 2) {
+      ASSERT_OK(replayer->Prepare());
+    }
+    Status s = Status::OK();
+    uint64_t last_ts = replayer->GetHeaderTimestamp();
+    int num_ops = 0;
+    // Looping until trace end.
+    while (s.ok()) {
+      std::unique_ptr<TraceRecord> record;
+      s = replayer->NextTraceRecord(&record);
+      // Skip unsupported operations.
+      if (s.IsNotSupported()) {
+        continue;
+      }
+      if (s.ok()) {
+        // A TraceRecord's timestamp must be no smaller than the header
+        // timestamp and the timestamp of its previous trace record.
```

Comment:
Since we are reading these from the file, won't this always be true?  

Shouldn't this condition be checked inside of Execute?

And shouldn't there be a test where this condition is not true to validate the check works?

Reply:
`s.ok()` is false when reaching the trace end, so the execution should only be called if `NextTraceRecord` returns OK.

The timestamp checking is trivial in this test. But I wonder if it's possible to have incorrect/out-of-order timestamps in a trace file?

I can add a test for manual replay, and move the current manual replay part into a separate test.

Reply:
Added `DBTest2.TraceAndManualReplay`.